### PR TITLE
chacha20poly1305: Initial implementation of AEAD cipher.

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds the chacha20poly1305 AEAD symmetric key cipher algorithm to RIOT-os. 

I didn't reuse the existing generic chacha code. There are some issues with that code that prevent reuse here. While on one hand it is more flexible by providing a configurable number of rounds, it requires a 8 byte nonce and only works on message sizes that are a multiple of 64 bytes. I plan on merging both implementations at some point though.

### Testing procedure

Execute the unittest on your favorite platforms and verify that the test doesn't fail. Furthermore check whether the documentation is sufficient for correct use.

### Issues/PRs references

depends on #9236, #10218 and #10219

### CODE REVIEW

**Comment by bergzand** [comment]
Implementation is around 1.8KB on the samr21-xpro.

**Comment by bergzand** [comment]
Rebased now that #9236 is merged

**Comment by bergzand** [comment]
Hello, thank you for your interest in RIOT-os and in this PR. The current status here is that it is waiting for a maintainer to review the code change. 

@jia200x You reviewed and merged #9236 right? Are you willing to review this one too?

**Comment by bergzand** [comment]
> @jia200x You reviewed and merged #9236 right? Are you willing to review this one too?

Sorry, that was @MrKevinWeiss, not @jia200x

**Comment by MrKevinWeiss** [comment]
I can give it a try on Monday!

**Comment by bergzand** [comment]
Thanks!

File: sys/include/crypto/chacha20poly1305.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#endif
+
+#define CHACHA20POLY1305_KEY_BYTES      (32U) /**< Key length in bytes */
+#define CHACHA20POLY1305_NONCE_BYTES    (12U) /**< Nonce length in bytes */
+#define CHACHA20POLY1305_TAG_BYTES      (16U) /**< Tag length in bytes */
+
+/**
+ * @brief Chacha20poly1305 state struct
+ */
+typedef struct
```
Comment (MrKevinWeiss) [inline sys/include/crypto/chacha20poly1305.h]: hmm I forgot about this...
Maybe change it to:
```c
typedef union
{
    /* We need both the state matrix and the poly1305 state, but nearly not at
     * the same time. This works as long as the first 8 members of state
     * overlap fully or completely not with the first and second key parts
     * from the @ref poly1305_ctx_t struct */

    uint32_t state[16];  /**< The current state of the key stream. */
    poly1305_ctx_t poly; /**< Poly1305 state for the MAC */

} chacha20poly1305_ctx_t;
```
To make it c99 compliant.

**Comment by aabadie** [comment]
> This won't be mergeable until after the release though since I think we are in feature freeze.

To avoid confusion, we are in *soft* feature freeze until next monday: this means no major things can be merged into master until *hard* feature freeze. What is a major change is left to the decision of the maintainer. IMHO, I don't think this PR can be considered as major.

**Comment by MrKevinWeiss** [comment]
Ahh than GO GO GO!

**Comment by aabadie** [comment]
To avoid even more confusion, we are in soft feature freeze until next ~monday~ friday (the 18th).

File: sys/crypto/chacha20poly1305.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+{
+    *a += *b;
+    uint32_t tmp = *a ^ *d;
+    *d = (tmp << c) | (tmp >> (32 - c));
+}
+
+void _add_initial(chacha20poly1305_ctx_t *ctx, const uint8_t *key,
+                  const uint8_t *nonce, uint32_t blk)
+{
+    for (unsigned i = 0; i < 4; i++) {
```
[the remark this answers is not in the document]
Comment (bergzand) [reply sys/crypto/chacha20poly1305.c]: > Is it standard to use just unsigned? Would it be better to specify unsigned int or just size_t?

IMHO just a personal preference, I don't mind changing it to `unsigned int`.

> Also maybe you could unroll the loop (I don't know if that can help with compiler optimizations in M4s)?

I'd really like to leave this as an exercise to the reader if you don't mind :)

[the remark this answers is not in the document]
Comment (bergzand) [reply sys/crypto/chacha20poly1305.c]: > Is it standard to use just unsigned? Would it be better to specify unsigned int or just size_t?

IMHO just a personal preference, I don't mind changing it to `unsigned int`.

> Also maybe you could unroll the loop (I don't know if that can help with compiler optimizations in M4s)?

I'd really like to leave this as an exercise to the reader if you don't mind :)

File: sys/include/crypto/chacha20poly1305.h (hunk 1 above)
Comment (bergzand) [reply sys/include/crypto/chacha20poly1305.h]: Fixed, thanks!

**Comment by MrKevinWeiss** [comment]
Squash please!

File: sys/crypto/chacha20poly1305.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (MrKevinWeiss) [reply sys/crypto/chacha20poly1305.c]: sure thing!

[the remark this answers is not in the document]
Comment (MrKevinWeiss) [reply sys/crypto/chacha20poly1305.c]: Sure thing!

**Pull request merged**